README.md
# spatialWeights

[![Build Status](https://travis-ci.org/christophergandrud/spatialWeights.svg?branch=master)](https://travis-ci.org/christophergandrud/spatialWeights)

This R package is designed to implement some of the spatial weights for time series cross sectional data discussed in 
[Neumayer and Plumper (2010)](http://eprints.lse.ac.uk/30750/1/Making%20spatial%20analysis%20operational(lsero).pdf).

Weights can be found in parallel for improved speed.

**NOTE:** I am working on it largely for a specific research project and do not (currently) intend to get it into shape for a CRAN release. I make it available here in case you are interested and want to build on it/find errors that I may have created.


NEWS
# Changes will be documented here

## Version 0.1.2

- **Breaking Changes**: `morans_i` argument syntax.

- `morans_i` argument now accepts `none`, `message`, and `table` to specify if
and how to output p-values from Moran's I.

## Version 0.1.1

- **Breaking Changes**: changes argument syntax.

- Find distances using Euclidean method by default with the ability to use 
others.

- Ability to print p-value from Moran's I Spatial Autocorrelation index.


R/monadic_spatial_weights.R
#' Find monadic spatial weights for a cross-sectional time-series data set
#' @param df a data frame containing the unit ID variable and time variables as
#' well as location' and dependent variables.
#' @param id_var a character string identifying the unit ID variable in
#' \code{df}.
#' @param time_var a character string identifying the time variable in
#' \code{df}.
#' @param location_var a character string identifying the location of the units
#' in \code{df}. This is used to create the weighting matrix. Note that the
#' function finds the relative distance between the units by subtracting their
#' 'location'.
#' @param y_var a character string identifying the dependent variable in
#' \code{df}. Note that an independent variable could also be supplied.
#' @param location_var_type character string allowing you to hard code
#' the measurement type of the location variable and thus how to determine the
#' weighting method Can be either \code{continuous} or \code{categorical}
#' If not specified, the type with be determined automatically.
#' @param weight_name character string providing a custom weighting variable
#' name.
#' @param method the distance measure to be used. Only relevant for numeric
#' location variables. This must be one of
#' \code{"euclidean"}, \code{"maximum"}, \code{"manhattan"}, \code{"canberra"},
#' \code{"binary"} or \code{"minkowski"}. Any unambiguous substring can be
#' given. See \code{\link{dist}} for details.
#' @param mc_cores The number of cores to use, i.e. at most how many child
#' processes will be run simultaneously. The option is initialized from
#' environment variable \code{MC_CORES} if set. Must be at least one, and
#' parallelization requires at least two cores.
#' @param na_rm logical whether or not to remove missing values.
#' @param morans_i character. Whether to print the p-value of Moran's I
#' Autocorrelation Index to the console (\code{message}), return only a table of
#' p-values (\code{table}), or \code{none}. Note only relevant for continuous
#' \code{y_var}. Join count analysis for categorical data not yet implemented.
#' @param ... arguments to pass to methods.
#'
#' @details Finds spatial effects in monadic data. See Neumayer and Plumper
#'(2010, 591) for details.
#'
#' The weights are found for each unit \eqn{i} given other units \eqn{k} at
#' time \eqn{t} using \eqn{\sum_{k}w_{kit}y_{kt}}.
#'
#' For continuous \code{location_var} \eqn{w} is the euclidean distance.
#' You can choose an alternate method with the \code{method} argument. See
#' \code{\link{dist}} for details.
#'
#' For factor/character \code{location_var} \eqn{w} is 1 if they share the same
#' location, 0 otherwise. The weight is further found by averaging over the
#' number of units at \eqn{t} that share the location of \eqn{i}.
#'
#' @return A data frame with three columns, one each for \code{id_var},
#' \code{time_var}, and the newly created spatial weights.
#'
#' @examples
#' # Create fake time series data
#' faked <- expand.grid(ID = letters, year = 2010:2015)
#' faked$located_continuous <- nrow(faked):1
#' faked$located_categorical <- sample(c('E', 'S', 'W', 'N'),
#'                                  nrow(faked), replace = TRUE)
#' faked$y <- nrow(faked):1 - 200
#'
#' # Find weights for continuous data
#' df_weights_cont <- monadic_spatial_weights(df = faked, id_var = 'ID',
#'                                          time_var = 'year',
#'                                          location_var = 'located_continuous',
#'                                          y_var = 'y', mc_cores = 1)
#'
#' # Find weights using for character data
#' df_weights_cat <- monadic_spatial_weights(df = faked, id_var = 'ID',
#'                                          time_var = 'year',
#'                                          location_var = 'located_categorical',
#'                                          y_var = 'y', mc_cores = 1)
#'
#' # Return a table of p-values from Moran's I spatial autocorrelation test statistic
#' moran_i_table <- monadic_spatial_weights(df = faked, id_var = 'ID',
#'                                          time_var = 'year',
#'                                          location_var = 'located_categorical',
#'                                          y_var = 'y', mc_cores = 1,
#'                                          morans_i = 'table')
#'
#' @source Neumayer, Eric, and Thomas Plumper. "Making spatial analysis operational:
#' commands for generating spatial effect variables in monadic and dyadic data."
#' Stata Journal 10.4 (2010): 585-605.
#'  \url{http://eprints.lse.ac.uk/30750/1/Making\%20spatial\%20analysis\%20operational(lsero).pdf}.
#'
#' @importFrom parallel mclapply
#' @importFrom dplyr bind_rows
#'
#' @export

monadic_spatial_weights <- function(df, id_var, time_var, location_var, y_var,
                                    location_var_type,
                                    weight_name,
                                    method = 'euclidean',
                                    mc_cores = 1, na_rm = TRUE,
                                    morans_i = 'message',
                                    ...)
{
    temp <- NULL

    morans_i <- tolower(morans_i)
    if (!(morans_i %in% c('none', 'message', 'table')))
        stop("morans_i must be one of 'none', 'message', or 'table'.", call. = FALSE)

    if (morans_i != 'none' & mc_cores > 1)
        morans_i <- 'none'
        message("Note: p-value of Moran's I is only found when mc_cores = 1.\n")

    if (missing(weight_name))
        weight_name <- sprintf('sp_wght_%s_%s', location_var, y_var)

    if (!missing(location_var_type)) {
        if (location_var_type == 'continuous') type_cont <- TRUE
        else if (location_var_type == 'factor') type_cont <- FALSE
        else stop(message("location_var_type can only be 'continuous' or 'categorical'."),
                    call. = FALSE)
    }
    if (missing(location_var_type)) {
        # Determine weighting scheme and inform user
        if (is.numeric(df[, location_var])) {
            message(sprintf(
                'Continuous location variable detected. Proximity found using method = %s.\n',
                method))
            type_cont <- TRUE
        }
        else if (is.character(df[, location_var]) |
                 is.factor((df[, location_var]))) {
            message('Categorical location detected.\nProximity found using Xi == Xk and group averaging.\n')
            if (method != 'euclidean')
                message('method argument ignored for categorical location variables')
            type_cont <- FALSE
        }
    }

    # Remove missing values
    if (na_rm) df <- DropNA(df, c(id_var, time_var, location_var, y_var))

    # Split by time period
    time_split <- split(df, f = df[, time_var])

    # Find all weights for each time period
    weighted <- mclapply(time_split, weights_at_t,
                   id_var = id_var,
                   location_var = location_var,
                   time_var = time_var,
                   weight_name = weight_name,
                   y_var = y_var,
                   type_cont = type_cont,
                   mc.cores = mc_cores,
                   method = method,
                   morans_i = morans_i)
    weighted <- suppressWarnings(bind_rows(weighted, .id = time_var))

    if (morans_i != 'table') {
        names(weighted) <- c(time_var, names(weighted)[-1])
        weighted <- weighted[, c(id_var, time_var, weight_name)]
    }
    return(weighted)
}


R/utils.R
#' Function to find monadic weights at one time point
#' @param df a data frame containing the unit ID variable and time variables as
#' well as location' and dependent variables.
#' @param id_var a character string identifying the unit ID variable in
#' \code{df}.
#' @param time_var a character string identifying the time variable in
#' \code{df}.
#' @param location_var a character string identifying the location of the units
#' in \code{df}. This is used to create the weighting matrix. Note that the
#' function finds the relative distance between the units by subtracting their
#' 'location'.
#' @param y_var a character string identifying the dependent variable in
#' \code{df}. Note that an independent variable could also be supplied.
#' @param type_cont whether the locations variables are continuous.
#' @param method the distance measure to be used. Only relevant for numeric
#' location variables. This must be one of
#' \code{"euclidean"}, \code{"maximum"}, \code{"manhattan"}, \code{"canberra"},
#' \code{"binary"} or \code{"minkowski"}. Any unambiguous substring can be
#' given. See \code{\link{dist}} for details.
#' @param return_matrix logical. Whether or not to return the adjacency matrix.
#' Could be useful for debugging.
#' @param weight_name character string providing a custom weighting variable
#' name.
#' @param morans_i character specifying whether to print the p-value of
#' Moran's I Autocorrelation Index to the console (\code{message}), return only
#' a table of p-values (\code{table}), or \code{none}.
#' @param ... arguments to pass to methods.
#'
#' @source Neumayer, Eric, and Thomas Plumper. "Making spatial analysis
#' operational: commands for generating spatial effect variables in monadic and
#' dyadic data." Stata Journal 10.4 (2010): 585-605.
#'  \url{http://eprints.lse.ac.uk/30750/1/Making\%20spatial\%20analysis\%20operational(lsero).pdf}.
#'
#' @importFrom stats dist
#' @importFrom dplyr %>% full_join select
#' @importFrom igraph graph_from_data_frame as_adjacency_matrix
#' @importFrom ape Moran.I
#'
#' @noRd
weights_at_t <- function(df, id_var, location_var, y_var, type_cont,
                         time_var,
                         method = 'euclidean', return_matrix = FALSE,
                         weight_name, morans_i = 'message', ...)
{
    freq <- NULL

    if (missing(weight_name)) weight_name <- sprintf('sp_wght_%s_%s',
                                                     location_var, y_var)

    if (missing(type_cont)) stop(
        'type_cont must be specified as TRUE or FALSE.', call. = FALSE)

    if (any(duplicated(df[, id_var]))) stop('Duplicate observations found',
                                            call. = FALSE)

    # Find w_{ikt}
    if (type_cont) {
        t_matrix <- as.matrix(dist(df[, location_var], method = method, ...))
    }
    else if (!isTRUE(type_cont)) {
        df$temp <- 1
        joined <- full_join(df, df, by = 'temp')
        joined <- joined[, c(paste0(id_var, '.y'), paste0(id_var, '.x'),
                             paste0(location_var, '.y'),
                             paste0(location_var, '.x'))]
        joined$weighting[joined[, 3] == joined[, 4]] <- 1
        joined$weighting[joined[, 3] != joined[, 4]] <- 0

        joined <- joined[, c(paste0(id_var, '.x'),
                             paste0(id_var, '.y'), 'weighting')]
        joined <- joined[joined[, 1] != joined[, 2], ]
        grph <- graph_from_data_frame(joined, directed = FALSE, vertices = NULL)
        t_matrix <- as_adjacency_matrix(grph, attr = 'weighting',
                                        sparse = FALSE)
    }

    if (return_matrix) return(t_matrix)
    else {
        # Find y_{kt}
        dependent_y <- df[, c(id_var, y_var)]

        if (morans_i != 'none') {
            # Find and print Moran's I
            if (type_cont) {
                t_matrix_inv <- 1/t_matrix
                diag(t_matrix_inv) <- 0
            }
            else t_matrix_inv <- t_matrix == 1

            mi <- Moran.I(dependent_y[, 2], t_matrix_inv)

            time_value <- unique(df[, time_var])
            mor_i <- round(mi$observed, digits = 3)
            m_p_value <- format.pval(mi$p.value, digits = 3)
        }


        if (morans_i == 'table') {
            out <- data.frame(morans_i = mor_i, morans_i_p_value = m_p_value)
            return(out)
        }

        else if (morans_i != 'table') {
            if (morans_i == 'message')
                message(sprintf("%s: Moran's I p-value: %s", time_value, m_p_value))

            matrix_product <- t_matrix * dependent_y[, 2]
            out <- colSums(matrix_product) %>% as.data.frame
            out[, id_var] <- df[, id_var]
            names(out) <- c(weight_name, id_var)

            if (!isTRUE(type_cont)) {
                # Find group averages
                counts <- table(df[, location_var]) %>% data.frame
                names(counts) <- c(location_var, 'freq')
                id_located <- merge(df, counts, by = location_var)
                id_located <- id_located[, c(id_var, location_var, 'freq')]
                out <- merge(out, id_located, by = id_var)
                out$freq <- out$freq - 1
                out$freq[out$freq < 1] <- 0
                out[, weight_name] <- out[, weight_name] / out$freq
                out <- out %>% select(-freq)
            }
            return(out)
        }
    }
}



#' Drop rows from a data frame with missing values on a given variable(s).
#' @source From the DataCombine package
#' @noRd

DropNA <- function(data, Var)
{
    # Find term number
    DataNames <- names(data)
    if (missing(Var)) {
        if (isTRUE(message)) {
            message('No Var specified. Dropping all NAs from the data frame.\n')
        }
        Var <- names(data)
    }
    TestExist <- Var %in% DataNames
    if (!all(TestExist)){
        stop("Variable(s) not found in the data frame.", call. = FALSE)
    }

    # Drop if NA
    if (length(Var) == 1){
        DataNoNA <- data[!is.na(data[, Var]), ]

        DataVar <- data[, Var]
        DataNA <- DataVar[is.na(DataVar)]
        TotalDropped <- length(DataNA)
    }
    else{
        RowNA <- apply(data[, Var], 1, function(x){any(is.na(x))})
        DataNoNA <- data[!RowNA, ]

        TotalDropped <- sum(RowNA)
    }

    if (TotalDropped > 0)
        message(paste(TotalDropped,
            "rows dropped from the data frame because of missing values.\n"))

    return(DataNoNA)
}
